File: windows/MonitorCore/ApiClient.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace CPAMP.MonitorCore;

public sealed class ApiClient : IDisposable
{
    private readonly HttpClient http;
    private readonly Uri baseUrl;
    public ApiClient(string url, string secret, HttpMessageHandler? handler = null)
    {
        baseUrl = Configuration.NormalizeUrl(url);
        http = new(handler ?? new HttpClientHandler { AllowAutoRedirect = false, UseCookies = false });
        http.Timeout = TimeSpan.FromSeconds(50);
        http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", secret);
        http.DefaultRequestHeaders.Accept.Add(new("application/json"));
    }
    public async Task<JsonObject> Request(string path, JsonObject? body, CancellationToken token)
    {
        using var request = new HttpRequestMessage(body is null ? HttpMethod.Get : HttpMethod.Post, new Uri(baseUrl, path));
        if (body is not null) request.Content = new StringContent(body.ToJsonString(), Encoding.UTF8, "application/json");
        using var response = await http.SendAsync(request, token);
        var status = (int)response.StatusCode;
        if (!response.IsSuccessStatusCode)
            throw new MonitorException(status switch {
                401 or 403 => "Authentication failed. Check the management key.",
                >= 300 and < 400 => "Redirect refused. Enter the final HTTPS service URL.",
                429 => "Rate limited. Retry later.", _ => $"Service request failed (HTTP {status})." }, status);
        return JsonNode.Parse(await response.Content.ReadAsStringAsync(token)) as JsonObject
            ?? throw new MonitorException("The service did not return a JSON object.");
    }
    public async Task<List<Account>> Accounts(CancellationToken token)
    {
        var data = await Request("v0/management/auth-files", null, token);
        if (data["files"] is not JsonArray) throw new MonitorException("Incompatible account list response.");
        return Json.Objects(data["files"]).Select(Account.FromJson).DistinctBy(a => a.Id)
            .OrderBy(a => a.Provider).ThenBy(a => a.Title).ToList();
    }
    public async Task<Dictionary<string, JsonObject>> History(List<Account> accounts, CancellationToken token)
    {
        Dictionary<string, JsonObject> result = [];
        foreach (var batch in accounts.Chunk(200))
        {
            var data = await Request("v0/management/monitoring/account-history",
                new() { ["accounts"] = new JsonArray(batch.Select(a => (JsonNode)a.Target()).ToArray()) }, token);
            if (data["items"] is not JsonArray) throw new MonitorException("Incompatible history response.");
            foreach (var item in Json.Objects(data["items"])) result[Json.Text(item["row_key"])] = item;
        }
        return result;
    }
    public async Task<QuotaResult> Quota(Account account, CancellationToken token)
    {
        if (account.AuthIndex.Length == 0) throw new MonitorException("The account has no auth_index.");
        var headers = new JsonObject { ["Authorization"] = "Bearer $TOKEN$", ["Content-Type"] = "application/json" };
        var body = new JsonObject { ["authIndex"] = account.AuthIndex, ["method"] = "GET" };
        switch (account.Provider)
        {
            case "codex":
                body["url"] = "https://chatgpt.com/backend-api/wham/usage";
                headers["User-Agent"] = "codex-tui/0.149.1";
                if (account.AccountId.Length > 0) headers["Chatgpt-Account-Id"] = account.AccountId;
                break;
            case "antigravity":
                body["url"] = "https://daily-cloudcode-pa.googleapis.com/v1internal:retrieveUserQuotaSummary";
                body["method"] = "POST";
                body["data"] = new JsonObject { ["project"] = account.ProjectId }.ToJsonString();
                headers["User-Agent"] = "antigravity/1.0.13";
                break;
            case "xai":
                body["url"] = "https://cli-chat-proxy.grok.com/v1/billing?format=credits";
                headers["x-xai-token-auth"] = "xai-grok-cli";
                headers["x-grok-client-version"] = "0.2.101";
                headers["User-Agent"] = "grok-pager/0.2.101 grok-shell/0.2.101 (windows; " +
                    System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture.ToString().ToLowerInvariant() + ")";
                break;
            case "claude":
                body["url"] = "https://api.anthropic.com/api/oauth/usage";
                headers["anthropic-beta"] = "oauth-2025-04-20";
                break;
            default: throw new MonitorException("Live quota is not supported for this provider.");
        }
        body["header"] = headers;
        var data = await Request("v0/management/api-call", body, token);
        var status = (int)(Json.Number(data["status_code"] ?? data["statusCode"]) ?? 0);
        if (status is < 200 or >= 300) throw new MonitorException(status == 429 ?
            "Quota rate limited. Retrying in 15 minutes." : $"Provider quota request failed (HTTP {status}).", status);
        var payload = data["body"] as JsonObject;
        if (payload is null && data["body"] is JsonValue raw && raw.TryGetValue<string>(out var text))
            payload = JsonNode.Parse(text) as JsonObject;
        return QuotaParser.Parse(account.Provider, payload ?? throw new MonitorException("Unrecognized quota response."), DateTimeOffset.UtcNow);
    }
    public static string SafeMessage(Exception error) => error switch {
        MonitorException e => e.Message,
        OperationCanceledException => "Request cancelled or timed out.",
        HttpRequestException => "Network connection failed. Check the service address and connection.",
        JsonException => "The response could not be parsed. Check the service version.",
        _ => "Operation failed. Check configuration and local storage permissions." };
    public void Dispose() => http.Dispose();
}

File: Sources/MonitorCore/APIClient.swift
import Foundation

private final class NoRedirects: NSObject, URLSessionTaskDelegate {
    func urlSession(_ session: URLSession, task: URLSessionTask,
                    willPerformHTTPRedirection response: HTTPURLResponse,
                    newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
        completionHandler(nil)
    }
}

public final class APIClient {
    private let base: URL
    private let key: String
    private let session: URLSession
    public init(baseURL: String, key: String) throws {
        base = try Configuration.normalizeURL(baseURL)
        self.key = key
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 35
        config.timeoutIntervalForResource = 50
        config.httpCookieStorage = nil
        config.urlCache = nil
        session = URLSession(configuration: config, delegate: NoRedirects(), delegateQueue: nil)
    }
    deinit { session.invalidateAndCancel() }

    public func request(_ path: String, body: [String: Any]? = nil) async throws -> [String: Any] {
        var req = URLRequest(url: base.appendingPathComponent(path))
        req.setValue("Bearer " + key, forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if let body {
            req.httpMethod = "POST"
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else { throw MonitorError("服务器响应无效。") }
        guard (200...299).contains(http.statusCode) else {
            let message: String
            switch http.statusCode {
            case 401, 403: message = "管理认证失败，请检查 CPAMP 管理密钥。"
            case 300...399: message = "服务发生重定向，请填写最终 HTTPS 地址。"
            case 429: message = "请求频率受限，稍后重试。"
            default: message = "CPAMP 请求失败（HTTP \(http.statusCode)）。"
            }
            throw MonitorError(message, status: http.statusCode)
        }
        guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw MonitorError("服务返回的不是 JSON 对象。")
        }
        return object
    }

    public func accounts() async throws -> [Account] {
        let data = try await request("v0/management/auth-files")
        guard let files = data["files"] as? [[String: Any]] else { throw MonitorError("账号列表结构不兼容。") }
        return files.map(Account.init).sorted { $0.provider == $1.provider ? $0.title < $1.title : $0.provider < $1.provider }
    }

    public func history(_ accounts: [Account]) async throws -> [String: [String: Any]] {
        var result: [String: [String: Any]] = [:]
        for offset in stride(from: 0, to: accounts.count, by: 200) {
            let batch = Array(accounts[offset..<min(offset + 200, accounts.count)])
            let data = try await request("v0/management/monitoring/account-history", body: ["accounts": batch.map(\.target)])
            for item in data["items"] as? [[String: Any]] ?? [] { result[string(item["row_key"])] = item }
        }
        return result
    }

    public func quota(_ account: Account) async throws -> QuotaResult {
        guard !account.authIndex.isEmpty else { throw MonitorError("缺少账号 auth_index。") }
        var headers = ["Authorization": "Bearer $TOKEN$", "Content-Type": "application/json"]
        var body: [String: Any] = ["authIndex": account.authIndex, "method": "GET"]
        switch account.provider {
        case "codex":
            body["url"] = "https://chatgpt.com/backend-api/wham/usage"
            headers["User-Agent"] = "codex-tui/0.149.1"
            if !account.accountID.isEmpty { headers["Chatgpt-Account-Id"] = account.accountID }
        case "antigravity":
            body["url"] = "https://daily-cloudcode-pa.googleapis.com/v1internal:retrieveUserQuotaSummary"
            body["method"] = "POST"
            body["data"] = String(data: try JSONSerialization.data(withJSONObject: ["project": account.projectID]), encoding: .utf8)!
            headers["User-Agent"] = "antigravity/1.0.13"
        case "xai":
            body["url"] = "https://cli-chat-proxy.grok.com/v1/billing?format=credits"
            headers["x-xai-token-auth"] = "xai-grok-cli"
            headers["x-grok-client-version"] = "0.2.101"
            headers["User-Agent"] = "grok-pager/0.2.101 grok-shell/0.2.101 (macos; aarch64)"
        case "claude":
            body["url"] = "https://api.anthropic.com/api/oauth/usage"
            headers["anthropic-beta"] = "oauth-2025-04-20"
        default: throw MonitorError("暂不支持此提供商的实时额度。")
        }
        body["header"] = headers
        let data = try await request("v0/management/api-call", body: body)
        let status = Int(number(data["status_code"] ?? data["statusCode"]) ?? 0)
        guard (200...299).contains(status) else {
            throw MonitorError(status == 429 ? "额度查询限流，15 分钟后重试。" : "提供商额度查询失败（HTTP \(status)）。", status: status)
        }
        let payload: [String: Any]?
        if let object = data["body"] as? [String: Any] { payload = object }
        else if let raw = (data["body"] as? String)?.data(using: .utf8) { payload = try JSONSerialization.jsonObject(with: raw) as? [String: Any] }
        else { payload = nil }
        guard let payload else { throw MonitorError("额度响应格式无法识别。") }
        return try QuotaParser.parse(provider: account.provider, payload: payload)
    }
}
